sources/api/pluto/src/aws.rs

```diff
@@ -0,0 +1,47 @@
+use aws_config::default_provider::credentials::DefaultCredentialsChain;
+use aws_config::imds;
+use aws_smithy_types::retry::{RetryConfig, RetryConfigBuilder};
+use aws_types::region::Region;
+use aws_types::SdkConfig;
+use snafu::{ResultExt, Snafu};
+use std::time::Duration;
+
+// Max request retry attempts
+const MAX_ATTEMPTS: u32 = 10;
```

Comment:
It looks like the random jitter applied to the backoff could cause all requests to be retried very quickly.

I'd like a stronger property, like "we'll retry for up to 3 minutes", without also getting into really long delays (tens of seconds), since then we might spend the last minute or two waiting for a retry that will never happen.

Reply:
It is kind of unfortunate that you can't set a maximum backoff in this SDK.

To achieve this property are you suggesting we write our own retry loop in pluto which combines with the internal retries to create the behavior we want?

Reply:
> as an overall timeout on the pluto execution, so we don't end up stacking lengthy timeouts together for each request that happens within an execution of the settings handle.

Double checking my understanding; Are we saying to wrap each `sundog` settings generator execution in a `tokio::time::timeout`?

> for the SDK client it's more of a "retry forever" strategy but without letting the backoff grow to absurd numbers.

Max backoff seems to cap out at 20 seconds: https://github.com/awslabs/aws-sdk-rust/blob/v0.15.0/sdk/aws-smithy-client/src/retry.rs#L95

Unfortunately there's no public interface that lets me customize that value: https://github.com/awslabs/aws-sdk-rust/blob/66423e05991ee831696bc32fe3e452694cf0d231/sdk/aws-smithy-client/src/retry.rs#L50-L51

However, we can change the multiplier for the exponential backoff. The default is 1 second and we can lower that if necessary: https://docs.rs/aws-sdk-ec2/latest/aws_sdk_ec2/config/retry/struct.RetryConfig.html#method.with_initial_backoff

In the worse case, the jitter can still cause each subsequent retry to happen almost immediately. So it's awkward to pick out an appropriate `MAX_ATTEMPTS` here. Assuming there's no backoff between retries. 3 minutes of retries would roughly equate to 58 retries with the default 3.1 second connection timeout.

So to summarize, I'll wrap each settings generator call in a 3 minute enforced tokio timeout block. Then I'll increase the max retry attempts for the SDK to ~60 (or higher) to accommodate the "worst" case scenario where backoffs are nonexistent due to jitters. To ensure the backoff doesn't grow to a ridiculous number, we scale the multiplier down to something like 100ms, so the backoff will essentially never grow beyond 100ms * 60 retries =  6 seconds. Does that sounds right?

Reply:
> So to summarize, I'll wrap each settings generator call in a 3 minute enforced tokio timeout block.
Sounds good. Maybe a 5 minute timeout for consistency with the IMDS client timeout.

> Then I'll increase the max retry attempts for the SDK to ~60 (or higher) to accommodate the "worst" case scenario where backoffs are nonexistent due to jitters. To ensure the backoff doesn't grow to a ridiculous number, we scale the multiplier down to something like 100ms, so the backoff will essentially never grow beyond 100ms * 60 retries = 6 seconds. Does that sounds right?

If the backoff is capped at 20 seconds, that might be OK too. Then we can just retry indefinitely within the time allotted by `sundog`.

Reply:
One problem I see with the retry-indefinitely strategy is for when we call `pluto cluster-dns-ip`. If the EKS describe cluster call fails for whatever reason, `pluto` defaults to a fallback value based on IMDS. In this new model,  we would never fallback and hard fail the boot, whereas before, if `pluto` couldn't reach the EKS endpoint, it would at least be able to set a value after the 5 min timeout in pluto.

Reply:
I opted to limit the settings generator execution to 6 mins in sundog and I kept the overall API request timeout in pluto to 5 minutes so as to give pluto a chance to use fallback values if possible.